compiler: introduce a mid-end IR (MIR)

zerbina opened the pull request:
Summary
=======

This PR introduces and integrates a **m**id-end **i**ntermediate
**r**epresentation (MIR) into the compiler. It is an IR for code and
meant to host target-independent transformations, analysis, and
lowerings, and is the first step towards implementing the plan detailed
in https://github.com/nim-works/nimskull/discussions/467.

Overview
--------

Before being processed by the code-generators, the `PNode` AST is
first passed to `mirgen`, which translates the AST into MIR code. In
the future, the MIR passes will take place after this step, but there
currently aren't any. After all eventual transformations are applied,
the MIR code is translated back into `PNode` AST. Rewriting the
code-generators to operate on the MIR directly is not feasible (at
least not as part of this commit), so the aforementioned translation is
required.

Until a proper pass management is implemented, `mirbridge` combines the
`mirgen`/`astgen` step into a single procedure, so that the logic can
be reused across all code-generators.

Passing the result of the MIR code -> `PNode`-AST translation to the
`injectdestructors` pass makes the latter produce code that fails the
current tests, so as a temporary workaround, the MIR processing is not
run at all for code for which the `injectdestructors` pass is enabled.

Lessons learned / Takeaways
===========================

`mirgen` has a similar structure to the other code-generators, in that
it also uses `genX` procedures that recursively call each other. It is
much more strict however, i.e. instead of using the top-level
dispatcher procedure everywhere, knowledge about the context is used in
order to call the correct procedure directly. For example, when
processing call arguments, the `gen` procedure for *expressions* is
called directly, as arguments are always (except for an explicitly
handled special-case) expressions. The consequences:
- more decisions happening at *compile-time* instead of at *run-time*.
  Decision making happens earlier -- it is moved outwards
- `mirgen` acts as a small validation layer for the input AST.
  Ill-formed or semantically invalid AST is (in some capacity) detected
  as a side-effect of how `mirgen` operates
- it is easier to reason about what AST forms are valid by only looking
  at the code (i.e. `mirgen` acts as a good overview of what is valid
  and what is not)

This helped uncovering a good amount of issues:
- the unsound collapsing of `HiddenAddr (HiddenDeref x)` node sequences
- `HiddenAddr` nodes having the wrong type in some cases
- when evaluating whether a `concept` applies and the last expression
  in the concept's body is the result of a template expansion, a
  `nkStmtListExpr` with expressions of which the result is not
  discarded is passed to `semConstExpr` (`vmgen` ignored this, but
  `mirgen` doesn't)
- when inserting boolean literals, `closureiters` used the symbol of
  the `bool` enum fields instead of integer literals. Past
  early-`transf` (where the `closureiters` transformation takes place)
  this is illegal, and only didn't fail code-generation because all
  code-generators still handled `skEnumField` symbols for robustness

Other findings:
- semantic analysis uses the VM to evaluate type expression in a lot of
  cases. For example, the expressions `static int` or `T not nil` are
  either partially or fully evaluated with the VM. The main work is
  done by `vmgen` here: an expression that has a `typeDesc` type is
  turned into a `nkType` node constant -- the VM only loads this
  constant into the result register and immediately exits again
- `vmgen` is being used by sem to detect and reject expressions
  containing unresolved generic parameters
- `closureiters` implements the same complex-expression-to-statement
  transformation as `mirgen`. Moving the `closureiters` transformation
  to operate on the MIR will likely simplify the logic there by a lot
- there are almost not tests for lifetime tracking hooks with the JS
  target
- cursor inference or the `injectdestructors` pass don't work with the
  output of `astgen`

MIR design
----------

What turned out to be a very good decision, was to separate the scoping
behaviour (in terms of both lifetime and visibility) from control-flow
constructs. That is, to make constructs like `if` (`mnkIf`) or `case`
(`mnkCase`) not affect the lifetime or visibility of definitions. This
simplified a lot of logic, made things clearer, and made important MIR
code possible that could not be represented before. The takeaway: if
there's no concrete reason to do so, don't combine/conflate unrelated
behaviour into a single construct.

Details
=======

This commit is for the most part self-contained, but multiple bugs and
issues in unrelated modules had to be fixed in order for `mirgen` to
work:
- `closureiters` now correctly emits `nkIntLit` nodes instead of
  `skEnumField` symbols for boolean literals
- `closureiters` now sets the type of a `case` expression to `nil`
  when transforming it into a statement
- cursor inference is separated from the `injectdestructors` pass.
  This is necessary in order for the MIR integration to happen
  in-between both
- prevent semantically invalid `nkStmtListExpr` AST from being
  evaluated with the VM during `concept` application

**General fixes:**
Fix `vmgen` raising an assertion for `x.len` where `x` was not directly
a string (e.g. a `sink string` or alias to `string`).

**Changes to `vmgen`:**
- since `mirgen` now handles type expressions, `vmgen` no longer has
  to. Only `nkType` (type literals) are code-gen'ed, everything else
  type related is rejected (an error is raised)
- the special handling for compile-time-only call arguments is
  removed - `mirgen` handles this now
- magics for which the instruction is selected based on the symbol
  name use exhaustive case statements now. This prevents magic symbol
  having an unexpected name from going unnoticed

**Other additions:**
- add a facility (`OptIndex`) for the ad-hoc creation of nilable
  versions of index-like types to the `dod_helpers` module
- add a simple half-open slice implementation to the `idioms` module
- add branch-related iterators to `ast_query`

**Added modules:**
- `mirtrees` contains the type definitions that make up the MIR, along
  with basic tree traversal routines
- `mirchangesets` implements the `Changeset` API, which is the primary
  way of modifying MIR code
- `mirconstr` contains some routines useful when generating MIR
  constructs. It's in an early stage of development, and the goal is to
  eventually generalize the mini DSL used by `mirgen` and move it here
- `sourcemaps` contains the `SourceMap` data type together with the
  outines for querying it. A `SourceMap` associates each `MirNode` with
  the `PNode` it originated from
- `mirgen` implements the `PNode` AST to MIR code translation
- `astgen` implements the MIR code to `PNode` AST translation
- `mir/utils` contains non-essential routines that are mainly meant to
  be used for inspection during debugging
- `utils/containers` implements various `seq`-based containers that are
  generally useful in DOD contexts

Misc
----

- implement printing the symbol kind in `astrepr`

Future directions
=================

The MIR processing is integrated into the C and JS backend the same way
as `transf` is: the code-generators have to explicitly invoke
both. When the body of a procedure or a top-level statement is
encountered:
1. `transf` is executed for the code
2. the transformed code is passed to the MIR processing step (if used)
3. the AST resulting from the previous step is passed on to the actual
   code-generation logic

This approach has the following problems:
- the code-generators need to know about the existence of both `transf`
  and the MIR processing step
- the logic for invoking both steps is duplicated into each backend
- should the `lambdalifting` pass become a MIR pass, the pass is likely
  going to require access to the MIR code of multiple procedures at
  once. With the current processing pipeline, the required state and
  logic for managing would, to some degree, also need to be duplicated
  into each backend/code-generator

Instead, using an approach similar to the one used by the VM backend
(`vmbackend`) for both the C and JS backend would make sense.
`vmbackend` is responsible for: figuring out which procedure to run
code-generation for, transforming the code (`transf`), applying the
MIR processing, and then passing the resulting AST to the
code-generator (`vmgen`). This could be generalized into a reusable
abstraction that is then used across all backends, removing the
duplication of logic and making backend processing more unified.

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

### To-dos
- [x] integrate the to-and-from translation step into:
  - [x] `cgen`
  - [x] `vmgen`
  - [x] `jsgen`
- [x] implement the things still missing from `mirgen` 
- [x] refactor, cleanup, and document the added modules
- [x] write a proper commit/PR message

zerbina commented:
The translation step is now fully integrated into the C and JS code-generators. For `vmgen`, it is partially integrated through `vmjit` (meaning that the VM back-end doesn't make use of it yet).

As a debugging aid and to help one visualize how the MIR code looks and how the translation affects the AST reaching the code-generator, I added three defines that work similar to `--expandArc`:
- `--define:nimShowMirInput=x`: echoes the `PNode` AST for the procedure/module with name `x` that is translated into MIR code. This is the AST the code-generator would see with the MIR translation step disabled.
- `--define:nimShowMir=x`: echoes the generated MIR code for the procedure/module with the name `x`
- `--define:nimShowMirOutput=x`: echoes the `PNode` AST that is generated from the MIR code for the procedure/module with the name `x`. This is the AST the code-generator operates on

The traces don't use proper diagnostic types yet, but are instead directly written out via `msgs.writeln`. In addition, the names of the defines are not final (feel free to suggest better ones) and they're also not documented in `debug.rst` yet.

Note that due to how the code-generators are implemented, nothing will be echoed for procedures that are not part of the alive code.

---

The largest part of the test suite for the C target works, with most of the remaining failing test being related to `injectdestructors`, possibly because the AST now violates some assumption made by the pass (but there might also be other bugs left still).

If the `injectdestructors` related failures are really only because of AST differences, then I think it'd make sense to disable the MIR translation step for procedures and modules marked with `sfInjectDestructors` (i.e. those that require destructor injection) until #450 is merged.

I'm currently focusing on making the remaining test work, and, after that, will work on refactoring and documenting the code.

zerbina commented:
The functional part is done and the tests pass. All that remains is cleaning up the code and adding documentation, which, due to the amount of code, is likely going to take a bit of time.

So far, the impact on compile time for me is at an additional 1.5-2 seconds for a single bootstrapping iteration (when using `-d:release --gc:refc`). Peak memory consumption for a single iteration increased by ~90 MB. The by far largest amount of time is spent inside `astgen` (i.e. the MIR to `PNode` translation), which makes sense, as it performs a significant amount of allocations.

zerbina commented:
Even though I'm still in the process of cleaning up and documenting the code, the PR can already be considered ready for review.

In addition, I've pushed the draft of a high-level MIR overview. My goal here is to get it into a state where it provides a solid foundation for incrementally improving it as part of future PRs. There are multiple ways to help in this area, namely by:
1. proof reading the document
2. asking "why", "what", "how" questions based on the text in the document
3. looking at the output of `-d:nimShowMir` and asking "why", "what", "how" questions

The questions resulting from both 2. and 3. would greatly help me in figuring out what parts still need to be expanded upon.

alaviss commented on compiler/ast/ast_query.nim:
```diff
@@ -765,3 +765,21 @@ proc `[]`*(node: PNode, slice: NodeSliceName): seq[PNode] =
     of SliceAllArguments, SliceAllBranches: node.sons[1..^1]
     of SliceBranchExpressions: node.sons[0 .. ^2]
+
+iterator branches*(node: PNode): tuple[position: int, n: PNode] =
```
One of the few times I'd love to have `enumerate` :p

alaviss commented on doc/mir.rst:
```diff
@@ -0,0 +1,161 @@
+  primitives such as 'goto's back into 'if', 'block', and 'break' is fairly
+  involved and would also introduce additional overhead to the
+  back-translation
+- some transformation passes currently using the MIR are simplre to implement
```
```suggestion
- some transformation passes currently using the MIR are simple to implement
```

alaviss commented on doc/debug.rst:
```diff
@@ -420,6 +420,29 @@ procedure.
+general, one can print the input and output to the MIR canonicalization step
+plus the corresponding `PNode`-AST output.
+
+To print the `PNode`-AST that reaches `mirgen`, `--define:nimShowMirInput=name`
```
Nothing to change here, but at one point we should create a kind of annotation you can tag on procs you want to debug (ie. a pragma) and maybe move away from using defines.

alaviss commented on compiler/utils/cursors.nim:
```diff
@@ -0,0 +1,48 @@
+##              destructor-based management scheme has no benefit in many cases
+
+type
+  Cursor*[T] = distinct T
```
Maybe this should be called `Shallow`. `Cursor` makes me think of pointers, which is not what this module is.

zerbina replied:
I called it `Cursor` because of the functional similarity to the `.cursor` pragma, and because the type is used in places where the pragma should've been used instead (but couldn't, due to the mentioned limitations).

If wanted, I can rename the type to `Shallow`, but I don't know what name to use for the module in that case.

alaviss replied:
I think it's fine to keep the name.

But I can see that it's too easy to edit an object shallow copied in `Cursor`, which can have devastating effects, for example: de-synchronized length between a seq and its cursor.

Given how unsafe the API is, I would want a kind of safety net (ie. preventing mutating ops) or something that screams unsafe.

Personally I would like to /not/ have this around. I would take a `ref seq` any day of the week just for the ownership semantics, and it's just a lot more foolproof.

zerbina replied:
The `Cursor` type together with the module and all usages is now removed. For `astgen`, I resorted to consuming the tree and source-map in order to get around the copies, but `Changeset` has to perform a full copy for now.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+
+  HOslice[T] = object
+    ## Half-open slice
+    a, b: T
```
Should be moved out elsewhere. This is duplicated on `mirchangesets` as well.

Also, what utility does this provide compared to `Slice` (ie. why couldn't `Slice` be used instead)?

zerbina replied:
Yep, I know and agree, but I don't know where to. A dedicated module feels off, same as with as a general utility one. The `idioms` module might be a good place to move it to.

As for its utility: its easier to use with unsigned integers (e.g. no special handling is required for empty slices) compared to `Slice`.

alaviss replied:
`utils`, `idioms`, it can be anywhere really. I'm more concerned with not duplicating it.

And this could potentially find a place in `system`, though we might want a different name then :P

zerbina replied:
I moved the `HOslice` type plus the associated routines that are generally useful to the `utils/idioms` module for now.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+## expressions, ignoring declarative constructs that are not relevant to nor
+## representable with the MIR.
+##
+## None of the MIR operations that imply structured control-flow produce a
+## value, so the input expression that do (``if``, ``case``, ``block``, and
+## ``try``) are translated into statements where the last expression in each
+## clause is turned into an assignment to, depending on the context where
+## they're used, either a temporary or existing lvalue expression. The latter
+## are forwarded to the generation procedures via ``Destination``.
```
Could be helpful to provide samples of "regular" AST and its corresponding MIR.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+    else:
+      result = n
+
+  var n {.cursor.} = n
```
I don't think you should use `{.cursor.}` here. The pragma's semantics is a mess and I'd prefer that we don't touch it if it's not explicitly required.

zerbina replied:
Cursor inference fails here, so the explicit annotation makes sense. Ignoring the semantics when used for object fields, what about the semantics do you consider to be a mess?

alaviss replied:
`{.cursor.}` documentation is very vague on how it works. It is said to "prevents object construction/destruction pairs" and the only behavior demonstrated is eliding all ref count ops.

It's an optimization that's potentially unsafe (ie. the object being viewed is destroyed), so I'd want to leave it out if it is not necessary, but then we have atomic RC in nimskull rn, so avoiding them is a pretty big thing...

zerbina replied:
The documentation is indeed lacking, but how the `.cursor` pragma works is straightforward: local variables annotated with it are non-owning locations for which all lifetime hooks are disabled. Assigning to cursor locations always uses a bit-wise copy.

For object fields, the pragma only works reliably for `ref` types and closures, but I consider this to be a design issue that can be easily fixed.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+    of nkPragmaBlock:
+      result = n.lastSon
+    of nkStmtListExpr:
+      # see if all nodes except the last one are empty. If they are, the
+      # ``nkStmtListExpr`` is skipped
```
Is this a `transf` thing? I'm not a fan of this kind of processing. Macros and `sem` can and do produce nested empty `StmtList`, so this is not the best way to filter through them.

You should document what kind of AST are you expecting so that we know when can this logic be deprecated.

zerbina replied:
I'm not sure I understand what you mean. The idea here is to skip/ignore redundant `StmtListExpr` nodes such as `StmtListExpr (Empty), someExpr` and get the underlying `someExpr`.

alaviss replied:
What I wanted to know is /how/ these nodes come to life. An `StmtListExpr` that have a bunch of `Empty` is pretty weird, no?

And what if that `StmtListExpr` is like this: `StmtListExpr(StmtListExpr() StmtListExpr() Call(...)`? Would `transf` or whatever that comes before collapse the inner empty exprs?

zerbina replied:
`sem` transforms `when` statements that don't apply into `Empty` nodes -- same with both `runnableExamples` calls and, during template expansion, `CommentStmt` nodes.

Neither empty `StmtListExpr` nor `StmtList` should get through semantic analysis (they currently do, but that's a separate issue).

Anyhow, in order to not duplicate the check and also handle `CommentStmt`s, I replaced the `anyIt` usage with `ast_query.stupidListExpr`.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+      # see if all nodes except the last one are empty. If they are, the
+      # ``nkStmtListExpr`` is skipped
+      for i in 0..<n.len-1:
+        if n[i].kind != nkEmpty:
```
`sequtils.anyIt` is probably a good fit here

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+  if isNimvm: n[0][1]
+  else:       n[1][0]
+
+func selectDefKind(s: PSym): range[mnkDef..mnkDefCursor] =
```
You probably don't want a `range` return type.

This means whatever variable you created from calling this function is tagged with that `range`. If you reuse that variable for other purposes, it might trigger a range error.

zerbina replied:
I do want a `range` return type here, both for documentation purposes and so that one doesn't have to use an `else: unreachable()` branch when using the result with a `case`.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+  x.kind != dkNone
+
+func initDestination(n: PNode, isFirst: bool): Destination =
+  result = Destination(kind: dkGen, node: n, isEmpty: isFirst, own: not isCursor(n))
```
```suggestion
  Destination(kind: dkGen, node: n, isEmpty: isFirst, own: not isCursor(n))
```

When it's an expression you don't have to use `result`

zerbina replied:
I know, but I'm still undecided whether I prefer using an explicit `result` assignment in such cases or not.

Does there exist a general preference?

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+    isEmpty: bool ## whether the destination has no value. Used to decide if
+                  ## an ``mnkInit`` or ``mnkAsgn`` is to be used
+    own: bool     ## whether the destination wants ownership over the assigned
+                  ## value. Used to select between a fast and normal assignment
```
Consider using flags (enum + set) instead. They're easier to query and have a smaller memory footprint.

Bonus points for easy extension in the future.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+
+  CodeFragment = object
+    nodes: MirNodeSeq
+    source: seq[SourceIndex]
```
This can use some docs

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+  ## already represent a half-open slice
+  HOslice[T](a: x.a, b: x.b)
+
+template toOpenArray[T, I](x: seq[T], slice: Slice[I]): openArray[T] =
+  toOpenArray(x, slice.a.int, slice.b.int)
+
+template toOpenArray[T, I](x: seq[T], slice: HOslice[I]): openArray[T] =
+  toOpenArray(x, slice.a.int, slice.b.int - 1)
```
I highly recommend caching `slice` in a temporary to avoid multiple evaluations.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+func apply(frag: var CodeFragment, id: SourceIndex) =
+  ## Associates all nodes added since the previous call to ``apply`` with the
+  ## origin information identified by `id`
+  assert frag.nodes.len > frag.source.len
```
You should drop this assert. For something like this I consider doing nothing a perfectly valid scenario.

zerbina replied:
Calling `apply` when there's nothing to update hints at a logic or programming error somewhere, hence the assert. I would prefer to keep it (but can update the documentation).

alaviss replied:
I feel like that's a long shot, given how `applySource` just manually do the check instead.

But I don't have strong opinions on this, so feel free to do whatever.

alaviss commented on compiler/mir/mirgen.nim:
```diff
@@ -0,0 +1,2105 @@
+  inc c.numTemps
+
+func nextLabel(c: var TCtx): LabelId =
+  result = LabelId(c.numLabels + 1)
```
I'd recommend `high(LabelId)` as the invalid id if you need those instead of shifting.

zerbina replied:
But then the zero-representation is not the 'none'/invalid value, which is a property that I dislike and that can cause issues easily.

alaviss reviewed (commented):
Initial look. I'll review the rest tomorrow.

alaviss commented on compiler/mir/sourcemaps.nim:
```diff
@@ -0,0 +1,69 @@
+    #       implementation detail
+  SourceId* = distinct uint32
+    ## A ``SourceIndex`` with an explicit "nil" state. This is essentially an
+    ## optimized ``Option[SourceIndex]``
```
These two needs distinct names. I kept conflating `Id` and `Index` reading the code.

alaviss replied:
I feel like you might want to just commit to 1-based indices instead of having two types for translation :p

zerbina replied:
The old `SourceId` is now gone and replaced with usage of the new `OptIndex` container type.

alaviss commented on compiler/mir/mirchangesets.nim:
```diff
@@ -0,0 +1,332 @@
+## This module implements the ``Changeset`` API, which is the main way of
```
I'm not a fan of this.

This system basically means that our trees are too expensive to modify, and hint at a symptom that has to be handled.

On top of that, the change sets are potentially dangerous, as changes must be batched carefully and you cannot check the state of the tree before the changes are applied. This means MIR processing code have to be extremely careful about backtracking and batching, making the pipeline harder to modify.

alaviss commented on compiler/utils/containers.nim:
```diff
@@ -0,0 +1,50 @@
+  ## Appends a new item to the Store and returns the ID assigned to
+  ## it
+  x.data.add it
+  result = I(x.data.high)
```
There's a potential problem where `I` is smaller than `int`: `I` could overflow and there's no facility to verify that's not the case.

zerbina replied:
I added a `rangeCheck` that makes sure that the resulting index is representable by `I`.

alaviss reviewed (commented):
I'll need to comb through this a couple more passes (I haven't fully evaluated `astgen`, `mirgen`, `mirtrees`), but here are my current remarks.

zerbina commented:
@alaviss I've addressed the blockers you mentioned on Matrix (`Cursor` usage, missing range checks, `Id`/`Index` confusion), but am leaving the PR open so that you can take another look -- if you think it's good to go, feel free to merge it.

zerbina committed:
compiler: introduce a mid-end IR (MIR)

Summary
=======

This PR introduces and integrates a **m**id-end **i**ntermediate
**r**epresentation (MIR) into the compiler. It is an IR for code and
meant to host target-independent transformations, analysis, and
lowerings, and is the first step towards implementing the plan detailed
in https://github.com/nim-works/nimskull/discussions/467.

Overview
--------

Before being processed by the code-generators, the `PNode` AST is
first passed to `mirgen`, which translates the AST into MIR code. In
the future, the MIR passes will take place after this step, but there
currently aren't any. After all eventual transformations are applied,
the MIR code is translated back into `PNode` AST. Rewriting the
code-generators to operate on the MIR directly is not feasible (at
least not as part of this commit), so the aforementioned translation is
required.

Until a proper pass management is implemented, `mirbridge` combines the
`mirgen`/`astgen` step into a single procedure, so that the logic can
be reused across all code-generators.

Passing the result of the MIR code -> `PNode`-AST translation to the
`injectdestructors` pass makes the latter produce code that fails the
current tests, so as a temporary workaround, the MIR processing is not
run at all for code for which the `injectdestructors` pass is enabled.

Lessons learned / Takeaways
===========================

`mirgen` has a similar structure to the other code-generators, in that
it also uses `genX` procedures that recursively call each other. It is
much more strict however, i.e. instead of using the top-level
dispatcher procedure everywhere, knowledge about the context is used in
order to call the correct procedure directly. For example, when
processing call arguments, the `gen` procedure for *expressions* is
called directly, as arguments are always (except for an explicitly
handled special-case) expressions. The consequences:
- more decisions happening at *compile-time* instead of at *run-time*.
  Decision making happens earlier -- it is moved outwards
- `mirgen` acts as a small validation layer for the input AST.
  Ill-formed or semantically invalid AST is (in some capacity) detected
  as a side-effect of how `mirgen` operates
- it is easier to reason about what AST forms are valid by only looking
  at the code (i.e. `mirgen` acts as a good overview of what is valid
  and what is not)

This helped uncovering a good amount of issues:
- the unsound collapsing of `HiddenAddr (HiddenDeref x)` node sequences
- `HiddenAddr` nodes having the wrong type in some cases
- when evaluating whether a `concept` applies and the last expression
  in the concept's body is the result of a template expansion, a
  `nkStmtListExpr` with expressions of which the result is not
  discarded is passed to `semConstExpr` (`vmgen` ignored this, but
  `mirgen` doesn't)
- when inserting boolean literals, `closureiters` used the symbol of
  the `bool` enum fields instead of integer literals. Past
  early-`transf` (where the `closureiters` transformation takes place)
  this is illegal, and only didn't fail code-generation because all
  code-generators still handled `skEnumField` symbols for robustness

Other findings:
- semantic analysis uses the VM to evaluate type expression in a lot of
  cases. For example, the expressions `static int` or `T not nil` are
  either partially or fully evaluated with the VM. The main work is
  done by `vmgen` here: an expression that has a `typeDesc` type is
  turned into a `nkType` node constant -- the VM only loads this
  constant into the result register and immediately exits again
- `vmgen` is being used by sem to detect and reject expressions
  containing unresolved generic parameters
- `closureiters` implements the same complex-expression-to-statement
  transformation as `mirgen`. Moving the `closureiters` transformation
  to operate on the MIR will likely simplify the logic there by a lot
- there are almost not tests for lifetime tracking hooks with the JS
  target
- cursor inference or the `injectdestructors` pass don't work with the
  output of `astgen`

MIR design
----------

What turned out to be a very good decision, was to separate the scoping
behaviour (in terms of both lifetime and visibility) from control-flow
constructs. That is, to make constructs like `if` (`mnkIf`) or `case`
(`mnkCase`) not affect the lifetime or visibility of definitions. This
simplified a lot of logic, made things clearer, and made important MIR
code possible that could not be represented before. The takeaway: if
there's no concrete reason to do so, don't combine/conflate unrelated
behaviour into a single construct.

Details
=======

This commit is for the most part self-contained, but multiple bugs and
issues in unrelated modules had to be fixed in order for `mirgen` to
work:
- `closureiters` now correctly emits `nkIntLit` nodes instead of
  `skEnumField` symbols for boolean literals
- `closureiters` now sets the type of a `case` expression to `nil`
  when transforming it into a statement
- cursor inference is separated from the `injectdestructors` pass.
  This is necessary in order for the MIR integration to happen
  in-between both
- prevent semantically invalid `nkStmtListExpr` AST from being
  evaluated with the VM during `concept` application

**General fixes:**
Fix `vmgen` raising an assertion for `x.len` where `x` was not directly
a string (e.g. a `sink string` or alias to `string`).

**Changes to `vmgen`:**
- since `mirgen` now handles type expressions, `vmgen` no longer has
  to. Only `nkType` (type literals) are code-gen'ed, everything else
  type related is rejected (an error is raised)
- the special handling for compile-time-only call arguments is
  removed - `mirgen` handles this now
- magics for which the instruction is selected based on the symbol
  name use exhaustive case statements now. This prevents magic symbol
  having an unexpected name from going unnoticed

**Other additions:**
- add a facility (`OptIndex`) for the ad-hoc creation of nilable
  versions of index-like types to the `dod_helpers` module
- add a simple half-open slice implementation to the `idioms` module
- add branch-related iterators to `ast_query`

**Added modules:**
- `mirtrees` contains the type definitions that make up the MIR, along
  with basic tree traversal routines
- `mirchangesets` implements the `Changeset` API, which is the primary
  way of modifying MIR code
- `mirconstr` contains some routines useful when generating MIR
  constructs. It's in an early stage of development, and the goal is to
  eventually generalize the mini DSL used by `mirgen` and move it here
- `sourcemaps` contains the `SourceMap` data type together with the
  outines for querying it. A `SourceMap` associates each `MirNode` with
  the `PNode` it originated from
- `mirgen` implements the `PNode` AST to MIR code translation
- `astgen` implements the MIR code to `PNode` AST translation
- `mir/utils` contains non-essential routines that are mainly meant to
  be used for inspection during debugging
- `utils/containers` implements various `seq`-based containers that are
  generally useful in DOD contexts

Misc
----

- implement printing the symbol kind in `astrepr`

Future directions
=================

The MIR processing is integrated into the C and JS backend the same way
as `transf` is: the code-generators have to explicitly invoke
both. When the body of a procedure or a top-level statement is
encountered:
1. `transf` is executed for the code
2. the transformed code is passed to the MIR processing step (if used)
3. the AST resulting from the previous step is passed on to the actual
   code-generation logic

This approach has the following problems:
- the code-generators need to know about the existence of both `transf`
  and the MIR processing step
- the logic for invoking both steps is duplicated into each backend
- should the `lambdalifting` pass become a MIR pass, the pass is likely
  going to require access to the MIR code of multiple procedures at
  once. With the current processing pipeline, the required state and
  logic for managing would, to some degree, also need to be duplicated
  into each backend/code-generator

Instead, using an approach similar to the one used by the VM backend
(`vmbackend`) for both the C and JS backend would make sense.
`vmbackend` is responsible for: figuring out which procedure to run
code-generation for, transforming the code (`transf`), applying the
MIR processing, and then passing the resulting AST to the
code-generator (`vmgen`). This could be generalized into a reusable
abstraction that is then used across all backends, removing the
duplication of logic and making backend processing more unified.

zerbina commented:
I've noticed and fixed multiple problems that weren't caught by the test suite:
- `jsgen` was not invoking `canonicalize` for all input AST
- because of the separation of cursor inference from the `injectdestructors` pass and me forgetting to adjust `jsgen`, cursor inference wasn't performed anymore when using the JS target 
- the `injectdestructors` pass doesn't work with the output of `astgen`

The last one is likely a problem with the implementation of the `injectdestructors` pass, but because of a bug, the MIR processing step was effectively skipped for code being flagged for destructor injection, thus hiding the issue. The to-and-from MIR translation still took place -- the resulting AST was just not used.

As a quick fix, I officially disabled the MIR step for code that is flagged for destructor injection for now. This prevents turning further transformations into MIR passes (since not all code reaching the code-generator is passed to MIR processing first), but fortunately, the PR that turns the `injectdestructors` pass into a MIR pass (#450) is almost ready to be merged.

zerbina commented:
bors r+
